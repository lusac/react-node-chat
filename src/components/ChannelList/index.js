import React from 'react'
import PropTypes from 'prop-types'
import NewChannelModal from '../NewChannelModal'

export default class ChannelList extends React.Component {
  constructor() {
    super()
    this.state = {
      showModal: false
    }
  }

  selectChat(id) {
    this.props.socket.emit('join channel', id)
  }

  toggleModal() {
    this.setState({
      showModal: !this.state.showModal
    })
  }

  render() {
    return (
      <div className="channel-list">
        <div className="channel-list__header">
          Ol√° <strong className="channel-list__header__name">{this.props.username}</strong>!
        </div>

        <div className="channel-list__section">
          <div className="channel-list__section__title">
            <span>Canais</span>
            <span
              className="plus"
              onClick={this.toggleModal.bind(this)}>+</span>
          </div>

          {Object.keys(this.props.channels).map(id => {
            return (
              <span
                key={id}
                className={"channel-list__item " + (this.props.channel.id === id ? 'selected' : '')}
                onClick={this.selectChat.bind(this, id)}>
                  # {this.props.channels[id].name}</span>
            )
          })}
        </div>

        {this.state.showModal &&
          <NewChannelModal
            toggleModal={this.toggleModal.bind(this)}
            socket={this.props.socket} />
        }
      </div>
    )
  }
}

ChannelList.propTypes = {
  socket: PropTypes.object,
  channel: PropTypes.object,
  username: PropTypes.string,
  channels: PropTypes.object,
};