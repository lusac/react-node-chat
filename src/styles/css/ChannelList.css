.channel-list {
  display: grid;
  height: 100vh;
  background: #f2f2f2;
  grid-template-rows: 70px auto;
}
.channel-list__header {
  display: flex;
  align-items: center;
  justify-content: center;
  border-right: 1px solid #e1e1e1;
}
.channel-list__header__name {
  color: #00c876;
  max-width: 110px;
  overflow: hidden;
  font-weight: bold;
  margin: 0 0 0 5px;
  white-space: nowrap;
  text-overflow: ellipsis;
}
.channel-list__section {
  display: flex;
  overflow: auto;
  flex-direction: column;
}
.channel-list__section__title {
  padding: 18px;
  border-right: 1px solid #e1e1e1;
}
.channel-list__item {
  cursor: pointer;
  padding: 6px 18px;
  border-right: 1px solid #e1e1e1;
}
.channel-list__item:hover {
  background: #EFEFEF;
}
.channel-list__item.selected {
  background: #FFF;
  border-right: none;
}

/*# sourceMappingURL=ChannelList.css.map */
