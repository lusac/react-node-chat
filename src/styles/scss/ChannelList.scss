.channel-list {
  display: grid;
  height: 100vh;
  background: #f2f2f2;
  grid-template-rows: 30px auto;

  &__header {
    display: flex;
    align-items: center;
    justify-content: center;
    border-right: 1px solid #e1e1e1;
  }

  &__section {
    overflow: auto;
    display: flex;
    flex-direction: column;

    &__title {
      padding: 18px;
      border-right: 1px solid #e1e1e1;
    }
  }

  &__item {
    cursor: pointer;
    padding: 6px 18px;
    border-right: 1px solid #e1e1e1;

    &:hover {
      background: #EFEFEF;
    }

    &.selected {
      background: #FFF;
      border-right: none;
    }
  }
}