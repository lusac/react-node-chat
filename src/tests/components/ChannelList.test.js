/* global it describe expect beforeEach jest */
import React from 'react'
import { shallow, configure } from 'enzyme'
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16'
import ChannelList from '../../components/ChannelList';

configure({ adapter: new Adapter() })

describe('ChannelList component', () => {
  describe('snapshot', () => {
    it('empty list', () => {
      const channels = {}
      const renderedValue =  renderer.create(<ChannelList channels={channels} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })

    it('with channels', () => {
      const notifications = {}
      const channel = {id:2, name:'channel 2'}
      const channels = { 1:{id:1, name:'channel 1'}, 2:{id:2, name:'channel 2'}, 3:{id:3, name:'channel 3'}}
      const renderedValue =  renderer.create(<ChannelList notifications={notifications} channel={channel} channels={channels} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })

    it('with notifications', () => {
      const
        notifications = {1: 3, 2: 0, 3: 10},
        channel = {id:2, name:'channel 2'},
        channels = { 1:{id:1, name:'channel 1'}, 2:{id:2, name:'channel 2'}, 3:{id:3, name:'channel 3'}},
        renderedValue =  renderer.create(<ChannelList notifications={notifications} channel={channel} channels={channels} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })
  })

  describe('rendered', () => {
    let wrapper
    const
      notifications = {1: 3, 2: 0, 3: 10},
      channel = {id:2, name:'channel 2'},
      channels = { 1:{id:1, name:'channel 1'}, 2:{id:2, name:'channel 2'}, 3:{id:3, name:'channel 3'}}

    beforeEach(() => {
      wrapper = shallow(<ChannelList notifications={notifications} channel={channel} channels={channels} />)
    })

    it('correct label', () => {
      expect(wrapper.find('.channel-list__item__name').at(0).text()).toBe('# channel 1')
      expect(wrapper.find('.channel-list__item__name').at(1).text()).toBe('# channel 2')
      expect(wrapper.find('.channel-list__item__name').at(2).text()).toBe('# channel 3')
    })

    it('correct badges', () => {
      expect(wrapper.find('.badge').get(0).props.children).toBe(3)
      expect(wrapper.find('.badge').get(1).props.children).toBe(10)
    })
  })

  it('selectChannel method', () => {
    let
      socket = {
        emit: (a, b) => (a, b)
      },
      notifications = {1: 3, 2: 0, 3: 10},
      channel = {id:2, name:'channel 2'},
      channels = { 1:{id:1, name:'channel 1'}, 2:{id:2, name:'channel 2'}, 3:{id:3, name:'channel 3'}},
      wrapper = shallow(<ChannelList socket={socket} notifications={notifications} channel={channel} channels={channels} />)

    const spy = jest.spyOn(wrapper.instance().props.socket, 'emit')
    wrapper.instance().selectChannel(3)
    expect(spy).toHaveBeenCalledWith('get channel', 3)
  })

  it('toggleModal method', () => {
    let
      notifications = {1: 3, 2: 0, 3: 10},
      channel = {id:2, name:'channel 2'},
      channels = { 1:{id:1, name:'channel 1'}, 2:{id:2, name:'channel 2'}, 3:{id:3, name:'channel 3'}},
      wrapper = shallow(<ChannelList notifications={notifications} channel={channel} channels={channels} />)

    expect(wrapper.instance().state.showModal).toBeFalsy()
    wrapper.instance().toggleModal()
    expect(wrapper.instance().state.showModal).toBeTruthy()
  })
})