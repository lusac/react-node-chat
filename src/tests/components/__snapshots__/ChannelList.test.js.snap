// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelList component snapshot empty list 1`] = `
<div
  className="channel-list"
>
  <div
    className="channel-list__header"
  >
    Olá 
    <strong
      className="channel-list__header__name"
    />
    !
  </div>
  <div
    className="channel-list__section"
  >
    <div
      className="channel-list__section__title"
    >
      <span>
        Canais
      </span>
      <span
        className="plus"
        onClick={[Function]}
      >
        +
      </span>
    </div>
  </div>
</div>
`;

exports[`ChannelList component snapshot with channels 1`] = `
<div
  className="channel-list"
>
  <div
    className="channel-list__header"
  >
    Olá 
    <strong
      className="channel-list__header__name"
    />
    !
  </div>
  <div
    className="channel-list__section"
  >
    <div
      className="channel-list__section__title"
    >
      <span>
        Canais
      </span>
      <span
        className="plus"
        onClick={[Function]}
      >
        +
      </span>
    </div>
    <span
      className="channel-list__item "
      onClick={[Function]}
    >
      <span
        className="channel-list__name"
      >
        # 
        channel 1
      </span>
    </span>
    <span
      className="channel-list__item "
      onClick={[Function]}
    >
      <span
        className="channel-list__name"
      >
        # 
        channel 2
      </span>
    </span>
    <span
      className="channel-list__item "
      onClick={[Function]}
    >
      <span
        className="channel-list__name"
      >
        # 
        channel 3
      </span>
    </span>
  </div>
</div>
`;

exports[`ChannelList component snapshot with notifications 1`] = `
<div
  className="channel-list"
>
  <div
    className="channel-list__header"
  >
    Olá 
    <strong
      className="channel-list__header__name"
    />
    !
  </div>
  <div
    className="channel-list__section"
  >
    <div
      className="channel-list__section__title"
    >
      <span>
        Canais
      </span>
      <span
        className="plus"
        onClick={[Function]}
      >
        +
      </span>
    </div>
    <span
      className="channel-list__item "
      onClick={[Function]}
    >
      <span
        className="channel-list__name"
      >
        # 
        channel 1
      </span>
      <span
        className="badge"
      >
        3
      </span>
    </span>
    <span
      className="channel-list__item "
      onClick={[Function]}
    >
      <span
        className="channel-list__name"
      >
        # 
        channel 2
      </span>
    </span>
    <span
      className="channel-list__item "
      onClick={[Function]}
    >
      <span
        className="channel-list__name"
      >
        # 
        channel 3
      </span>
      <span
        className="badge"
      >
        10
      </span>
    </span>
  </div>
</div>
`;
